spec/core/ics-004-channel-and-packet-semantics/README.md

```diff
@@ -481,20 +540,122 @@ function chanCloseConfirm(
     connection = provableStore.get(connectionPath(channel.connectionHops[0]))
     abortTransactionUnless(connection !== null)
     abortTransactionUnless(connection.state === OPEN)
+
+    let counterpartyHops: []string = []
+    if (connectionHops.length > 1) {
+      counterpartyHops = getCounterPartyHops(proofInit)
+    } else {
+      counterpartyHops = [connection.counterpartyConnectionIdentifier]
+    }
+
     expected = ChannelEnd{CLOSED, channel.order, portIdentifier,
-                          channelIdentifier, [connection.counterpartyConnectionIdentifier], channel.version}
-    abortTransactionUnless(connection.verifyChannelState(
-      proofHeight,
-      proofInit,
-      channel.counterpartyPortIdentifier,
-      channel.counterpartyChannelIdentifier,
-      expected
-    ))
+                          channelIdentifier, counterpartyHops, channel.version}
+
+    if (connectionHops.length > 1) {
+      key = host.ChannelPath(counterparty.PortId, counterparty.ChannelId)
+
+      abortTransactionUnless(connection.verifyMultihopMembership(
+        connection,
+        proofHeight,
+        proofInit,
+        channel.connectionHops,
+        key
+        expected))
+    } else {
+      abortTransactionUnless(connection.verifyChannelState(
+        proofHeight,
+        proofInit,
+        channel.counterpartyPortIdentifier,
+        channel.counterpartyChannelIdentifier,
+        expected
+      ))
+    }
+
     channel.state = CLOSED
     provableStore.set(channelPath(portIdentifier, channelIdentifier), channel)
 }
 ```
 
+The `chanCloseFrozen` function is called by a relayer to force close a multi-hop channel if any client state in the
```

Comment:
Can we have a separate pseudocode on how the proofs are built? I think they will look differently depending to which channel end they are sent to. In the example A--B--xC--D (where x is the frozen client for B hosted on C), A cannot receive but could sent while D cannot sent but could receive.
I would expect then some logic below to determine if the client is for one of the connection hops or the counterparty of a connection hop. And basically find the C-B-A connection hops from the A--D channel connection hops when the close is received by A.

Reply:
If I understand your example, the frozen client proof would look like this:

1. Multi-hop consensus/connection proof from C-->A. This is the same as a regular multi-hop proof without the key/value proof.
2. key/value proof on `chain B`. For a frozen client state we need to know the key of the frozen client state on chainB, but this is actually tricky since we don't know the client ID. This is why we need the multi-hop proof in step 1 to include the connection state on the misbehaving chain C. The connectionEnd counterparty contains the client id on chain B we can use to derive the key for the client state on chain B. The client state value is complex and is provided with the proof.

I realize we never talked about the implementation details for frozen client proofs, but do you see any issues with this approach? It is efficient in some sense because it allows re-using the same multi-hop proof struct with an additional `index` parameter for frozen clients and allows use of a single multi-hop proof. The `index` is used to generalize the multi-hop proof so that the key/value can be proven on any consensus state in the mutli-hop channel path (in this case the client state proven on chainB instead of chainC).

Although similar to standard multi-hop proving, it does have some distinct differences. I'll add pseudo code for this.

Reply:
Not sure I understand 2. If we look at the `chanCloseFrozen` sent to A 
So for the case `A--B--xC--D`: The connection proof from 1 should have the B-end of the connection B--C. Once verified, we can get the client ID on C (for the B--C connection) from the counterparty of the B-end. That should match the client for the key/value proof.
In contrast let's take the case `A--Bx--C--D`: In this case the key/value proof should be for the client of the B-end of the connection and not the counterparty.

Reply:
Whoops, sorry, my example didn't follow your original case - I was assuming that `chainC` was the misbehaving chain with the frozen client on `chainB`.

Yes, the way you explained it is exactly how I was thinking. I'll push the pseudo code shortly.

Reply:
done

Reply:
do we have a pseudocode on how the proofs are constructed?